### ISSUE TITLE
Support custom function spaces in LFRic (multi-data fields, different number of layers and data)

### ISSUE THREAD

**Issue by TeranIvy** [stfc/PSyclone#868]
~~Relevant LFRic documentation is in [this LFRic wiki](https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/FieldEnhancements/MultidataPsycloneMetaProposal).~~

**Update**: The support will require a redesign of function space representation. See [this new LFRic wiki](https://code.metoffice.gov.uk/trac/lfric/wiki/LFRicInfrastructure/FieldEnhancements/FunctionSpaceMetadata).

**Update, 21/10/2025**: The design was discussed with STFC and the idea is to not have custom function spaces to capture the changes in `nlayers` and `ndata` as they are properties of the field and not function space continuity (the design of LFRic stores dofmaps in function space object). The current conclusion is to have additional metadata in the field argument types.

**Comment by TeranIvy** [comment]
- [ ] Metadata parsing support and related documentation,
- [ ] Kernel arguments for `k`-first fields (declarations, call argument list, stub argument list),
- [ ] Kernel arguments for `i`-first fields (declarations, call argument list, stub argument list),
- [ ] Kernel arguments for operators (declarations, call argument list, stub argument list),
- [ ] Kernel argument lists for basis functions/quadrature (declarations, call argument list, stub argument list).

**Comment by arporter** [comment]
Having discussed with @TeranIvy, we have agreed that we will need some recognised names for numbers of layers because there are physics 'kernels' with ~100 arguments but only two or three different values of nlayers - we don't want to have to extend the argument list as that would be massive.

**Comment by arporter** [comment]
Therefore, we need an agreed set of names that will be added to argument_mod.F90 and recognised by PSyclone.

**Comment by arporter** [comment]
@sergisiso has pointed out that `LFRicKernelConstTrans` could make use of the nlayers metadata rather than relying on a user providing a value (provided of course that the metadata specifies a value rather than a name). Note that that transformation is still in `psyclone.transformations` so needs to be moved first.

The transformation could also be extended to know about `ndata`.

**Comment by arporter** [comment]
Hi @tommbendall @TeranIvy and @stevemullerworth, is the idea that this new metadata would apply to operator arguments too?

**Comment by stevemullerworth** [comment]
> Hi [@tommbendall](https://github.com/tommbendall) [@TeranIvy](https://github.com/TeranIvy) and [@stevemullerworth](https://github.com/stevemullerworth), is the idea that this new metadata would apply to operator arguments too?

I would have thought there isn't a strong requirement as the requirement for custom function spaces is mainly finite difference code that doesn't use operators or the available function space mathematics.

**Comment by arporter** [comment]
Hi @stevemullerworth and @tommbendall, now that the metadata support is on, I'm beginning the implementation of the code generation. As usual, that has raised more questions. As an example, for the following kernel metadata:
```fortran
  type, extends(kernel_type) :: testkern_nlayers_ndata_type
     type(arg_type), dimension(7) :: meta_args =                           &
          (/ arg_type(gh_scalar, gh_real, gh_read),                        &
             arg_type(gh_field,  gh_real, gh_inc,  w1),                    &
             arg_type(gh_field,  gh_real, gh_read, w2),                    &
             ! Non-default number of layers.
             arg_type(gh_field,  gh_real, gh_read, w2, nlayers="shallow"), &
             ! Non-default number of layers but same as previous arg. so
             ! has same dof map.
             arg_type(gh_field,  gh_real, gh_read, w2, nlayers="shallow"), &
             ! Non-default number of data values per dof.
             arg_type(gh_field,  gh_real, gh_read, w2, ndata="precip"),    &
             ! Although we have seen the values of ndata and nlayers before
             ! (individually), the dof map will be different.
             arg_type(gh_field,  gh_real, gh_read, w2, ndata="precip",     &
                      nlayers="shallow")                                   &
           /)
     integer :: operates_on = cell_column
   contains
     procedure, nopass :: code => testkern_nlayers_ndata_code
  end type testkern_nlayers_ndata_type
```
I'm currently working towards a subroutine interface like:
```fortran
  subroutine testkern_nlayers_ndata_code(&
       nlayers, ascalar, fld1, fld2,     &
       fld3, nlayers_fld3,               &
       fld5,                             &
       fld4, ndata_fld4,                 &
       fld6,                             &
       ndf_w1, undf_w1, map_w1,          &
       ndf_w2, undf_w2, map_w2,          &
       ndf_fld3, undf_fld3, map_w2_fld3, &
       ndf_fld4, undf_fld4, map_w2_fld4, &
       ndf_fld6, undf_fld6, map_w2_fld6)
```
My question is: is it necessary to pass the `ndata` and `nlayers` values into the kernel subroutine? Presumably the field data arrays are still dimensioned by the appropriate `ndf`? If it is necessary/desirable, are you happy with the way I've chosen to order arguments above? i.e. each field that has a custom `nlayers` and/or `ndata` that hasn't already been seen is followed by arguments that hold those values (and will be looked-up in the PSy layer). So, for example, although `fld6` does have both a custom `nlayers` *and* `ndata`, those values have already been passed for earlier arguments so aren't repeated.

**Comment by stevemullerworth** [comment]
Certainly, it makes sense and is consistent to declare each unique value once, and in the location relating to where it is first required.

It may be more logical to align them with the `ndf, undf, map` arguments. When checking for correctness, I think one would check the field order (which currently matches the metadata) then check the map order. In the latter process, one currently needs to think about which duplicate definitions can be removed. Adding `ndata` and `nlayers` here means all the thinking about duplication is in a distinct place.

To your second query, `nlayers` and `ndata` values would always be needed. `ndf` declares only one of the "array" dimensions: all `W3` fields of a given order have the same `ndf` (so `ndf` is always 1 for all lowest order fields) regardless of the value of `nlayers` or `ndata`.

**Comment by tommbendall** [comment]
I also agree about the declaration of each unique value once. I had originally agreed with Steve that it would make more sense to pair this data with the other function space data, i.e. for example would be:
```
  subroutine testkern_nlayers_ndata_code(              &
       nlayers, ascalar, fld1, fld2,                   &
       fld3,                                           &
       fld5,                                           &
       fld4,                                           &
       fld6,                                           &
       ndf_w1, undf_w1, map_w1,                        &
       ndf_w2, undf_w2, map_w2,                        &
       ndf_fld3, undf_fld3, map_w2_fld3, nlayers_fld3, &
       ndf_fld4, undf_fld4, map_w2_fld4, ndata_fld4,   &
       ndf_fld6, undf_fld6, map_w2_fld6)
```
But on typing it out, it pairing the new `nlayers` and `ndata` arguments the `fld3` and `fld4` arguments might be obscuring that `fld5` and `fld6` also use these values.

So I think I might prefer grouping all extra `nlayers` and `ndata` arguments. I would put them first to group them with the existing `nlayers` argument:
```
  subroutine testkern_nlayers_ndata_code(&
       nlayers,                          &
       nlayers_fld3,                     &  ! All extra nlayer values
       ndata_fld4,                       &  ! All extra ndata values
       ascalar, fld1, fld2,              &
       fld3,                             &
       fld5,                             &
       fld4,                             &
       fld6,                             &
       ndf_w1, undf_w1, map_w1,          &
       ndf_w2, undf_w2, map_w2,          &
       ndf_fld3, undf_fld3, map_w2_fld3, &
       ndf_fld4, undf_fld4, map_w2_fld4, &
       ndf_fld6, undf_fld6, map_w2_fld6  &
)
```

PS: I was also not sure if `fld4` and `fld5` are labelled correctly in this example?

**Comment by arporter** [comment]
> PS: I was also not sure if `fld4` and `fld5` are labelled correctly in this example?

Don't worry - I think that was an artefact of the way the test kernel evolved.

**Comment by stevemullerworth** [comment]
Tom's alternative seems reasonable, though I didn't see why it helped explain that the values applied to `fld5` and `fld6`

Was there a reason why you can't have `nlayers_shallow` etc. ?

**Comment by arporter** [comment]
> Was there a reason why you can't have `nlayers_shallow` etc. ?

No. In fact that's how they are currently named in the PSy layer. Of course the kernel author can name the subroutine arguments any way they wish :-)

**Comment by tommbendall** [comment]
> Tom's alternative seems reasonable, though I didn't see why it helped explain that the values applied to `fld5` and `fld6`
> 
> Was there a reason why you can't have `nlayers_shallow` etc. ?

Yes if the user names them `nlayers_shallow` etc that would be clearer!

My thinking was that by separating the nlayers and ndata arguments from ndf/undf/map, we avoid a developer thinking the ndata/nlayer argument is only relevant to the first field

**Comment by stevemullerworth** [comment]
Sorry, I was being dim about the naming! If Tom's suggestion is doable, I support it.
